File: leetcode/js/find-maximum-number-of-string-pairs.js
/**
 * @param {string[]} words
 * @return {number}
 */
var maximumNumberOfStringPairs = function (words) {
  let result = 0;
  const set = new Set();

  for (const word of words) {
    if (set.has(word)) {
      result++;
    } else {
      set.add([...word].reverse().join(""));
    }
  }

  return result;
};

File: leetcode/python/find-maximum-number-of-string-pairs.py
class Solution:
    def maximumNumberOfStringPairs(self, words: list[str]) -> int:
        result = 0
        word_set = set()

        for word in words:
            if word[::-1] in word_set:
                result += 1
            
            word_set.add(word)
        
        return result